testonly/hammer/replay.go

```diff
@@ -0,0 +1,227 @@
+	if err == nil {
+		if msg, ok := reply.(proto.Message); ok {
+			ri.dumpMessage(msg)
+		} else {
+			glog.Warningf("failed to convert response %T to proto.Message", req)
+		}
+	}
+	return err
+}
+
+func (ri *recordingInterceptor) dumpMessage(in proto.Message) {
+	ri.mu.Lock()
+	defer ri.mu.Unlock()
+	if err := writeMessage(ri.outLog, in); err != nil {
+		glog.Error(err.Error())
+	}
+}
+
+func writeMessage(w io.Writer, in proto.Message) error {
+	a, err := ptypes.MarshalAny(in)
+	if err != nil {
+		return fmt.Errorf("failed to marshal %T %+v to any.Any: %v", in, in, err)
+	}
+	data, err := proto.Marshal(a)
+	if err != nil {
+		return fmt.Errorf("failed to marshal any.Any: %v", err)
+	}
+	// Encode as [4-byte big-endian length, message]
+	lenData := make([]byte, 4)
+	binary.BigEndian.PutUint32(lenData, uint32(len(data)))
+	w.Write(lenData)
+	w.Write(data)
+	return nil
+}
+
+func readMessage(r io.Reader) (*any.Any, error) {
+	// Decode from [4-byte big-endian length, message]
+	var lenData [4]byte
+	dlen, err := r.Read(lenData[:])
+	if err == io.EOF {
+		return nil, io.EOF
+	}
+	if dlen < 4 {
+		return nil, fmt.Errorf("corrupt data: expected 4-byte length, only found %d bytes", dlen)
+	}
+	l := binary.BigEndian.Uint32(lenData[:])
+	data := make([]byte, l)
+	n, err := r.Read(data)
+	if uint32(n) < l {
+		return nil, fmt.Errorf("corrupt data: expected %d bytes of data, only found %d bytes", n, l)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to read %d bytes of data: %v", l, err)
+	}
+	var a any.Any
+	if err := proto.Unmarshal(data, &a); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal into any.Any: %v", err)
+	}
+	return &a, nil
+}
+
+// ReplayFile reads recorded gRPC requests and re-issues them using the given
+// client.  If a request has a MapId field, and its value is present in mapmap,
+// then the MapId field is replaced before replay.
+func ReplayFile(ctx context.Context, r io.Reader, cl trillian.TrillianMapClient, mapmap map[int64]int64) {
+	for {
+		a, err := readMessage(r)
+		if err != nil {
+			if err != io.EOF {
+				glog.Errorf("Error reading message: %v", err)
+			}
+			return
+		}
+		glog.V(2).Infof("Replay %q", a.TypeUrl)
+		replayMessage(ctx, cl, a, mapmap)
+	}
+}
+
+// convertMessage modifies msg in-place so that the contents of a "MapId" field
+// are updated according to mapmap.
+func convertMessage(msg proto.Message, mapmap map[int64]int64) {
+	// Look for a "MapId" field that we can overwrite if needed.
+	pVal := reflect.ValueOf(msg)
+	if fieldVal := pVal.Elem().FieldByName("MapId"); fieldVal.CanSet() {
+		from := fieldVal.Int()
+		if to, ok := mapmap[from]; ok {
+			glog.V(2).Infof("Replacing msg.MapId=%d with %d in %T", from, to, msg)
+			fieldVal.SetInt(to)
+		}
+	}
+}
+
+func replayMessage(ctx context.Context, cl trillian.TrillianMapClient, a *any.Any, mapmap map[int64]int64) error {
+	switch a.TypeUrl {
+	case "type.googleapis.com/trillian.GetMapLeavesRequest":
```

Comment:
If we use a proto any, doesn't this type switch become unnecessary?

Reply:
Good point, a lot of it can be removed (I'd missed the existence of `DynamicAny`).

However, I can't see a way to get a concrete response type (nor to map from the request type to the relevant method on the `trillian.TrillianMapClient`), so there's still a (smaller) type switch.